### ISSUE TITLE
Secure Enclave for SSH Key Management

### ISSUE THREAD

**Issue by imlibra** [blinksh/blink#457]
Secure Enclave is a hardware security engine in iDevices, which is able to handle ECC keypairs. The keypairs in Secure Enclave will never be exported and considered be secure.

There are a dozens of Demo Apps for it on GitHub:
[SecureEnclaveCrypto](https://github.com/trailofbits/SecureEnclaveCrypto)
[EllipticCurveKeyPair](https://github.com/agens-no/EllipticCurveKeyPair)
......
[Apple Documents](https://developer.apple.com/documentation/security/certificate_key_and_trust_services/keys/storing_keys_in_the_secure_enclave)

**Comment by carloscabanero** [comment]
I don’t know if I understand you correctly, but I think we already do this. Keychain services APIs use Secure Enclave as a way to store data at rest and we have been using that since first version. We use a wrapper around Keychain APIs called UICKeyChainStore.

Closing here but let me know if you would like to see something else.

**Comment by omor1** [comment]
I think what's being asked here is to _generate_ the private key entirely within the Secure Enclave. A similar concept is available for desktop using [SeKey](https://github.com/ntrippar/sekey) as an SSH Agent.

**Comment by yury** [comment]
Hi @omor1.

‘config’ -> ‘+’ -> ‘Create new’ will generate the private key and store it in Secure Enclave.

**Comment by thegushi** [comment]
@yury -- I think what's being asked for is that the key be kept ENTIRELY in the secure enclave, which would make it non-exportable.  Currently only ED25519 keys work for this.  You can also set a flag on the key that it *requires* biometry to use.

Right now, if I generate a key, it would seem I have an option to "copy private key".  If that option is a no-op, it should be grayed out.

https://medium.com/@alx.gridnev/ios-keychain-using-secure-enclave-stored-keys-8f7c81227f4

**Comment by imlibra** [comment]
@carloscabanero That’s exactly what I mean.

**Comment by thegushi** [comment]
Reading over this, I realize this wasn’t totally clear — this is a requirement to GENERATE the key in the secure enclave and mark it non-exportable.

The logic in doing so is that the key is forever tied *to your device*.  If you have three devices you’re using blink on, it’s three keys, rather than one private key you’ve copied to all the places.

If you lose/wipe your device, that key is effectively destroyed because it cannot be copied off (and also, cannot have been copied ON, since using things like SEkey on the mac requires that the SE be the generator).   

Even if you use your fingerprint, enroll someone else’s finger, etc, the MOST you can do is re-use the key.  You can’t obtain it.

Hope that makes more sense.

**Comment by pzeinlinger** [comment]
This would be great from a security point of view. Apparently, Termius supports it (https://docs.termius.com/termius-handbook/connecting-to-a-host/keychain/sep-secure-enclave-processor).

**Comment by imlibra** [comment]
Still no improve on this...

Secure Terminal also support SEKey now, but not blinksh.

**Comment by carloscabanero** [comment]
I think this is a nice to have, but unless someone is really blocked by it, very low priority. We currently already store the keys encrypted by the SE, and although it is true that signing happens in memory, given the nature of these devices the difference is minimal. 

We have other battles, but if someone wants to jump in, we can definitely guide the implementation.

**Comment by thegushi** [comment]
On Sat, 13 Feb 2021, carloscabanero wrote:

> 
> I think this is a nice to have, but unless someone is really blocked by it, very low priority. We currently already store the keys encrypted by the SE, and although it is true that
> signing happens in memory, given the nature of these devices the difference is minimal.

The difference is the guarantee that this key CAN ONLY be used with this 
device.  (Absent some exploit of the entire secure enclave mechanism).

I'm not a coder, just a sysadmin (who works on some very production 
critical stuff), but I'd call this nice to have from a "security policy" 
point of view.

-Dan

> We have other battles, but if someone wants to jump in, we can definitely guide the implementation.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.[AAIWKKG2INXOYKCWBKDP2VTS624YDA5CNFSM4ETO4772YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOFZUVCVI.gif]
> 
> 
>

-- 

"She's been getting attacked by these leeches, they're leaving these marks
all over her neck. You gotta keep her out of those woods.  If one more
leech gets her, she's gonna get a smack."

-Someone's Mother, December 18th, 1998

--------Dan Mahoney--------
Techie,  Sysadmin,  WebGeek
Gushi on efnet/undernet IRC
FB:  fb.com/DanielMahoneyIV
LI:   linkedin.com/in/gushi
Site:  http://www.gushi.org
---------------------------

**Comment by carloscabanero** [comment]
Happy to report, this is now a new feature on 13.90.3.273!!

**Comment by tommycusick** [comment]
Great news!  Is there a place on the GitHub project where I can request to join the TestFlight beta?  Eager to try it out.

**Comment by carloscabanero** [comment]
Send us a DM on Twitter or a message on Discord and we will get you rolling

**Comment by thegushi** [comment]
Ditto here, will send a DM shortly.
